lib/connectors/SentinelConnector/index.ts

```diff
@@ -269,19 +294,108 @@ export default class SentinelConnector extends AbstractConnector {
       commandTimeout: this.options.sentinelCommandTimeout,
       dropBufferSupport: true,
     });
+  }
+
+  private async resolve(
+    endpoint: Partial<ISentinelAddress>
+  ): Promise<ITcpConnectionOptions | null> {
+    const client = this.connectToSentinel(endpoint);
 
     // ignore the errors since resolve* methods will handle them
     client.on("error", noop);
 
+    let result: ITcpConnectionOptions | null = null;
+
     try {
       if (this.options.role === "slave") {
-        return await this.resolveSlave(client);
+        result = await this.resolveSlave(client);
       } else {
-        return await this.resolveMaster(client);
+        result = await this.resolveMaster(client);
       }
+
+      if (result) {
+        this.initFailoverDetector({
+          address: endpoint,
+          isConnected: true,
+          getClient: () => client,
+        });
+      }
+
+      return result;
     } finally {
-      client.disconnect();
+      if (!result) {
+        // Only disconnect if we didn't get a result.
+        // Otherwise we'll use this connection for failover detection.
+        client.disconnect();
+      }
+    }
+  }
+
+  private initFailoverDetector(firstSentinel: ISentinel): void {
+    // Move the current sentinel to the first position
+    this.sentinelIterator.reset(true);
+
+    const sentinels: ISentinel[] = [firstSentinel];
+
+    // Skip the first sentinel that we've already connected to
+    this.sentinelIterator.next();
+
+    // In case of a large amount of sentinels, limit the number of concurrent connections
+    while (sentinels.length < this.options.sentinelMaxConnections) {
+      const { done, value } = this.sentinelIterator.next();
+
+      if (done) {
+        break;
+      }
+
+      let client: IRedisClient | null = null;
+
+      const sentinel = {
+        address: value,
+        isConnected: false,
+        getClient: () => {
+          if (!client) {
+            client = this.connectToSentinel(value);
+            sentinel.isConnected = true;
+          }
+
+          return client;
+        },
+      };
+
+      sentinels.push(sentinel);
     }
+
+    this.sentinelIterator.reset(false);
+
+    if (this.failoverDetector) {
+      // Clean up previous detector
+      this.failoverDetector.cleanup();
+    }
+
+    this.failoverDetector = new FailoverDetector(this, sentinels);
+  }
+
+  private async activateFailoverDetector() {
+    if (!this.failoverDetector) {
+      return;
+    }
+
+    for (const client of this.failoverDetector.getClients()) {
+      // Apply reconnect strategy to sentinels now that we're no longer looking for master
+      client.options.retryStrategy = this.options.sentinelReconnectStrategy;
```

Comment:
Not too sure about changing options after the client has been initialised, could lead to something breaking in the future.

But I also couldn't think of a better solution at the moment.
During the initial master resolution, `retryStrategy` must be `null` so we can quickly skip unreachable sentinels.
A cleaner solution would be to disconnect and then reconnect with a new `retryStrategy`, but it seems wasteful not to reuse an already established connection.

Reply:
Yeah, the current implementation looks good to me.

> Not too sure about changing options after the client has been initialised

Generally speaking, we don't support modifying options after the client is initialized. Although most options should take effect and work as expected, there are no tests to cover that. In the next major version, I think we can make options read-only (or most likely just flag them as deprecated so users don't have to update their existing working code) and provide setters to change options that we support in an explicit way.

> It seems wasteful not to reuse an already established connection.

Connecting to Redis should be pretty lightweight so I think the right behavior and maintainability & readability of the code should take priority over reusing existing connections.